module/VuFind/src/VuFind/Navigation/AbstractMenu.php

```diff
@@ -38,53 +43,222 @@
+     *
+     * @var ?array
+     */
+    protected ?array $menu;
+
+    /**
+     * Constructor.
+     *
+     * @param array $sectionConfig Section configuration
+     */
+    public function __construct(array $sectionConfig)
+    {
+        $this->requiredSettings[self::GROUP_CONTEXT] ??= [];
+        $this->requiredSettings[self::ITEM_CONTEXT] ??= [];
+        $this->localizableSettings[self::GROUP_CONTEXT] ??= [];
+        $this->localizableSettings[self::ITEM_CONTEXT] ??= [];
+        $this->setSectionConfig($sectionConfig);
+    }
+
+    /**
+     * Set section configuration.
+     *
+     * @param array $sectionConfig Section configuration
+     *
+     * @return $this
+     */
+    public function setSectionConfig(array $sectionConfig): static
+    {
+        parent::setSectionConfig($sectionConfig);
+        $this->menu = null;
+        return $this;
+    }
+
+    /**
+     * Get section service.
+     *
+     * @return SectionServiceInterface
+     * @throws Exception if section service has not been set
+     */
+    public function getSectionService(): SectionServiceInterface
+    {
+        // Section service must be set after constructing the object. This
+        // requirement will be removed in VuFind version 12.
+        if (!isset($this->sectionService)) {
+            throw new Exception('Section service not set');
+        }
+        return $this->sectionService;
+    }
+
+    /**
+     * Set section service.
+     *
+     * This method must be called after constructing the object. This
+     * requirement will be removed in VuFind version 12.
+     *
+     * @param SectionServiceInterface $sectionService Section service
+     *
+     * @return static
+     */
+    public function setSectionService(SectionServiceInterface $sectionService): static
+    {
+        $this->sectionService = $sectionService;
+        return $this;
+    }
+
+    /**
+     * Localize section configuration.
+     *
+     * This method should be called after setting the section service and if
+     * setting the configuration outside the constructor. This requirement will
+     * be removed in VuFind version 12.
+     *
+     * @return static
+     */
+    public function localizeSectionConfig(): static
+    {
+        $config = $this->getSectionService()->localizeSettings($this);
+        foreach ($config as $group => $settings) {
+            $config[$group]
+                = $this->getSectionService()->localizeSettings(
+                    $this,
+                    $settings,
+                    self::GROUP_CONTEXT
+                );
+            foreach ($settings['MenuItems'] ?? [] as $i => $menuItem) {
+                $config[$group]['MenuItems'][$i]
+                    = $this->getSectionService()->localizeSettings(
+                        $this,
+                        $menuItem,
+                        self::ITEM_CONTEXT
+                    );
+            }
+        }
```

Comment:
Maybe worth using a variable to reduce getter calls; we can also reformat a little to improve readability:


```suggestion
        $sectionService = $this->getSectionService();
        $config = $sectionService->localizeSettings($this);
        foreach ($config as $group => $settings) {
            $config[$group] = $sectionService->localizeSettings(
                $this,
                $settings,
                self::GROUP_CONTEXT
            );
            foreach ($settings['MenuItems'] ?? [] as $i => $menuItem) {
                $config[$group]['MenuItems'][$i] = $sectionService->localizeSettings(
                    $this,
                    $menuItem,
                    self::ITEM_CONTEXT
                );
            }
        }
```